src/consensus/upgrades.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2018 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "consensus/upgrades.h"
+
+const struct NUInfo NetworkUpgradeInfo[Consensus::MAX_NETWORK_UPGRADES] = {
+    {
+        /*.nBranchId =*/ 0,
+        /*.strName =*/ "Sprout",
+        /*.strInfo =*/ "The Zcash network at launch",
+    },
+    {
+        /*.nBranchId =*/ 0x74736554,
+        /*.strName =*/ "Test dummy",
+        /*.strInfo =*/ "Test dummy info",
+    },
+    {
+        /*.nBranchId =*/ 0x5ba81b19,
+        /*.strName =*/ "Overwinter",
+        /*.strInfo =*/ "TBD",
+    }
+};
+
+UpgradeState NetworkUpgradeState(
+    int nHeight,
+    const Consensus::Params& params,
+    Consensus::UpgradeIndex idx)
+{
+    assert(nHeight >= 0);
+    assert(idx >= Consensus::UPGRADE_SPROUT && idx < Consensus::MAX_NETWORK_UPGRADES);
+    auto nActivationHeight = params.vUpgrades[idx].nActivationHeight;
+
+    if (nActivationHeight == Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT) {
+        return UPGRADE_DISABLED;
+    } else if (nHeight >= nActivationHeight) {
+        // From ZIP ???:
+        //
+        // ACTIVATION_HEIGHT
+        //     The block height at which the network upgrade rules will come into effect.
+        //
+        //     For removal of ambiguity, the block at height ACTIVATION_HEIGHT - 1 is
+        //     subject to the pre-upgrade consensus rules.
+        return UPGRADE_ACTIVE;
+    } else {
+        return UPGRADE_PENDING;
+    }
+}
+
+bool NetworkUpgradeActive(
+    int nHeight,
+    const Consensus::Params& params,
+    Consensus::UpgradeIndex idx)
+{
+    return NetworkUpgradeState(nHeight, params, idx) == UPGRADE_ACTIVE;
+}
+
+int CurrentEpoch(int nHeight, const Consensus::Params& params) {
+    for (auto idxInt = Consensus::MAX_NETWORK_UPGRADES - 1; idxInt >= Consensus::UPGRADE_SPROUT; idxInt--) {
+        if (NetworkUpgradeActive(nHeight, params, Consensus::UpgradeIndex(idxInt))) {
+            return idxInt;
+        }
+    }
+}
+
+uint32_t CurrentEpochBranchId(int nHeight, const Consensus::Params& params) {
+    return NetworkUpgradeInfo[CurrentEpoch(nHeight, params)].nBranchId;
+}
+
+bool IsActivationHeight(
+    int nHeight,
+    const Consensus::Params& params,
+    Consensus::UpgradeIndex idx)
+{
+    assert(idx >= Consensus::UPGRADE_SPROUT && idx < Consensus::MAX_NETWORK_UPGRADES);
+
+    // Don't count Sprout as an activation height
```

Comment:
Why not just `assert(idx > Consensus::UPGRADE_SPROUT && idx < Consensus::MAX_NETWORK_UPGRADES);` if not counting Sprout as activation height?

Reply:
That requires the caller to handle that case, and we know better here how that case should be handled (specifically, it's because Sprout wasn't actually an upgrade but is being included as one for code niceness, and this case is needed to prevent the entire chain being reindexed on first start due to missing, and unnecessary, index data).

Reply:
ACK, makes sense.